src/bin/snapshots.rs

```diff
@@ -0,0 +1,122 @@
+extern crate crossbeam;
+extern crate gll;
+extern crate insta;
+extern crate proc_macro2;
+extern crate rayon;
+extern crate rust_grammar;
+extern crate walkdir;
+extern crate regex;
+extern crate lazy_static;
+
+use {
+    insta::assert_snapshot_matches,
+    rust_grammar::parse,
+    std::{fmt::Debug, fs, process::exit},
+    walkdir::WalkDir,
+    regex::Regex,
+    lazy_static::lazy_static,
+};
+
+fn to_debug_str(debug: &dyn Debug) -> String {
+    format!("{:#?}", debug)
+}
+
+macro_rules! snapshot {
+    ($production:ident, $src:expr) => {
+        match $src.parse::<proc_macro2::TokenStream>() {
+            Ok(tts) => parse::$production::parse_with(tts, |_, result| to_debug_str(&result)),
+            // FIXME(eddyb) provide more information in this error case.
+            Err(_) => to_debug_str(&Err::<(), _>(parse::ParseError::<()>::NoParse)),
+        }
+    };
+}
+
+macro_rules! dispatch {
+    ($src:ident, $s:expr; $($prod:ident)*) => {
+        match $s {
+            $(stringify!($prod) => snapshot!($prod, $src),)*
+            prod => panic!("Unexpected production {} tested", prod),
+        }
+    };
+}
+
+lazy_static! {
+    static ref RE: Regex = Regex::new(r"\(\s+_,?\s+\)").unwrap();
+}
+
+fn test_snapshot(file: walkdir::DirEntry) {
+    let path = file.path();
+    let file_name = file.file_name().to_str().unwrap();
+    let src = fs::read_to_string(path).unwrap();
+    let production = &file_name[..file_name.find('.').unwrap_or(file_name.len())];
+    let forest = dispatch! { src, production;
+        // abi.lyg
+        Abi
+        // attr.lyg
+        OuterAttr InnerAttr Attr AttrInput
+        // expr.lyg
+        Expr ExprKind UnaryOp BinaryOp BinaryAssignOp FieldName StructExprFieldsAndBase
+        StructExprField StructExprFieldKind Label
+        If Cond ElseExpr MatchArm ClosureArg
+        // generics.lyg
+        Generics GenericParam GenericParamKind ForAllBinder WhereClause WhereBound LifetimeBound
+        TypeBound TypeTraitBound GenericArgs AngleBracketGenericArgsAndBindings GenericArg
+        TypeBinding
+        // item.lyg
+        ModuleContents Item ItemKind UseTree UseTreePrefix ForeignItem ForeignItemKind TraitItem
+        TraitItemKind ImplItem ImplItemKind FnHeader FnDecl FnArgs FnArg EnumVariant EnumVariantKind
+        StructBody TupleField RecordField
+        // macro.lyg
+        MacroCall MacroInput ItemMacroCall ItemMacroInput
+        // pat.lyg
+        Pat PatRangeValue Binding SlicePatElem TuplePatField StructPatFieldsAndEllipsis
+        StructPatField
+        // path.lyg
+        Path RelativePath PathSegment QSelf QPath
+        // stmt.lyg
+        Stmt Block
+        // type.lyg
+        Type FnSigInputs FnSigInput
+        // vis.lyg
+        Vis VisRestriction
+    };
+    let forest = forest.replace("Span..Span", "_");
+    let forest = RE.replace_all(&forest, "(_)");
```

Comment:
If you're processing the output like this, feel free to also replace `"_ => "` away.
And then maybe we can use `.snap.rs` as an extension so GitHub does syntax highlighting?
Although it might not be much.

Reply:
Actually, what we should do is [use `.gitattributes` to tell Linguist to treat `.snap` as Rust code](https://github.com/github/linguist#using-gitattributes).

If we change to using `serde::Serialize` (and RON, maybe?) we might be able to do more principled [redactions](https://docs.rs/insta/0.7.4/insta/#redactions). I'll definitely add the `_ => ` removal though. Is the `(_)` useful or should we just make them "unit structs"?

Reply:
> Is the `(_)` useful or should we just make them "unit structs"?

"Unit structs" is probably fine. As for serializing, the issue is we have things like `|` and the `[x, y, ...z]` spread syntax (in lists).